File: services/ecommerce-app-node/api-server/src/services/storage.js
const path = require('path');
const fs = require('fs');
const { v4: uuidv4 } = require('uuid');

const STORAGE_TYPE = process.env.STORAGE_TYPE || 'local';
const UPLOADS_DIR = path.join(__dirname, '../../uploads');

async function uploadFile(file) {
  if (STORAGE_TYPE === 's3') {
    return uploadToS3(file);
  }

  return uploadToLocal(file);
}

function uploadToLocal(file) {
  if (!fs.existsSync(UPLOADS_DIR)) {
    fs.mkdirSync(UPLOADS_DIR, { recursive: true });
  }

  const ext = path.extname(file.originalname);
  const fileName = `${uuidv4()}${ext}`;
  const filePath = path.join(UPLOADS_DIR, fileName);

  fs.writeFileSync(filePath, file.buffer);

  return `/uploads/${fileName}`;
}

async function uploadToS3(file) {
  const { getS3Client } = require('../config/aws');
  const { PutObjectCommand } = require('@aws-sdk/client-s3');

  const s3 = getS3Client();
  const ext = path.extname(file.originalname);
  const key = `uploads/${uuidv4()}${ext}`;

  const command = new PutObjectCommand({
    Bucket: process.env.S3_BUCKET,
    Key: key,
    Body: file.buffer,
    ContentType: file.mimetype,
    ACL: 'public-read',
  });

  await s3.send(command);

  const region = process.env.S3_REGION || 'ap-northeast-2';
  return `https://${process.env.S3_BUCKET}.s3.${region}.amazonaws.com/${key}`;
}

async function getPresignedUrl(fileName, fileType) {
  if (STORAGE_TYPE !== 's3') {
    throw new Error('Pre-signed URL can only be used when STORAGE_TYPE is s3');
  }

  const { getS3Client } = require('../config/aws');
  const { PutObjectCommand } = require('@aws-sdk/client-s3');
  const { getSignedUrl } = require('@aws-sdk/s3-request-presigner');

  const s3 = getS3Client();
  const ext = path.extname(fileName);
  const key = `uploads/${uuidv4()}${ext}`;

  const command = new PutObjectCommand({
    Bucket: process.env.S3_BUCKET,
    Key: key,
    ContentType: fileType,
    ACL: 'public-read',
  });

  const uploadUrl = await getSignedUrl(s3, command, { expiresIn: 900 });

  const region = process.env.S3_REGION || 'ap-northeast-2';
  const fileUrl = `https://${process.env.S3_BUCKET}.s3.${region}.amazonaws.com/${key}`;

  return { uploadUrl, fileUrl };
}

async function getPresignedGetUrl(url, expiresIn = 3600) {
  if (!url) return url;
  if (url.startsWith('/')) return url;
  if (url.includes('X-Amz-Signature')) return url;

  const bucket = process.env.S3_BUCKET;
  if (!bucket || !url.includes(`${bucket}.s3`)) return url;

  const { getS3Client } = require('../config/aws');
  const { GetObjectCommand } = require('@aws-sdk/client-s3');
  const { getSignedUrl } = require('@aws-sdk/s3-request-presigner');

  const urlObj = new URL(url);
  const key = urlObj.pathname.startsWith('/') ? urlObj.pathname.slice(1) : urlObj.pathname;

  const s3 = getS3Client();
  const command = new GetObjectCommand({
    Bucket: bucket,
    Key: key,
  });

  return getSignedUrl(s3, command, { expiresIn });
}

module.exports = { uploadFile, getPresignedUrl, getPresignedGetUrl };

File: services/ecommerce-app-fastapi/api-server-fastapi/app/services/storage.py
import uuid
from pathlib import Path

from app.config.settings import settings

UPLOADS_DIR = settings.UPLOADS_DIR


async def upload_file(file_data: bytes, filename: str, content_type: str) -> str:
    if settings.STORAGE_TYPE == "s3":
        return await _upload_to_s3(file_data, filename, content_type)

    return _upload_to_local(file_data, filename)


def _upload_to_local(file_data: bytes, filename: str) -> str:
    UPLOADS_DIR.mkdir(parents=True, exist_ok=True)

    ext = Path(filename).suffix
    new_filename = f"{uuid.uuid4()}{ext}"
    file_path = UPLOADS_DIR / new_filename

    with open(file_path, "wb") as file_handle:
        file_handle.write(file_data)

    return f"/uploads/{new_filename}"


async def _upload_to_s3(file_data: bytes, filename: str, content_type: str) -> str:
    from app.config.aws import get_s3_client

    s3 = get_s3_client()
    ext = Path(filename).suffix
    key = f"uploads/{uuid.uuid4()}{ext}"

    s3.put_object(
        Bucket=settings.S3_BUCKET,
        Key=key,
        Body=file_data,
        ContentType=content_type,
        ACL="public-read",
    )

    return f"https://{settings.S3_BUCKET}.s3.{settings.S3_REGION}.amazonaws.com/{key}"


async def get_presigned_url(file_name: str, file_type: str) -> dict:
    if settings.STORAGE_TYPE != "s3":
        raise ValueError("Pre-signed URL can only be used when STORAGE_TYPE is s3")

    from app.config.aws import get_s3_client

    s3 = get_s3_client()
    ext = Path(file_name).suffix
    key = f"uploads/{uuid.uuid4()}{ext}"

    upload_url = s3.generate_presigned_url(
        "put_object",
        Params={
            "Bucket": settings.S3_BUCKET,
            "Key": key,
            "ContentType": file_type,
            "ACL": "public-read",
        },
        ExpiresIn=900,
    )

    file_url = f"https://{settings.S3_BUCKET}.s3.{settings.S3_REGION}.amazonaws.com/{key}"

    return {"uploadUrl": upload_url, "fileUrl": file_url}
